lib/rules/no-unused-vars.js

```diff
@@ -832,6 +836,636 @@ module.exports = {
+                    // fix [[...[b]]]
+                    if (hasSingleElement(parentNode)) {
+                        if (parentNode.parent.type === "ArrayPattern") {
+                            return fixNestedArrayVariable(parentNode);
+                        }
+
+                        // fix 'const [...[b]] = foo; and function foo([...[b]]) {}
+                        return fixVariables(parentNode);
+                    }
+
+                    // fix [[a, ...[b]]]
+                    return fixer.removeRange([getPreviousTokenStart(node), node.range[1]]);
+                }
+
+                return null;
+            }
+
+            // skip fix when variable is reassigned
+            if (writeReferences.length > 1 && parentType !== "AssignmentPattern") {
+                return null;
+            }
+
+            // skip fix when initialized variable is reassigned
+            if (writeReferences.length > 0 && parent.init === null && parentType !== "AssignmentPattern") {
+                return null;
+            }
+
+            // skip fix when function parameters are ressaigned
+            if (isFunction(parent) && writeReferences.length > 0 && parent.params.includes(id)) {
+                return null;
+            }
+
+            // remove declared variables such as var a; or var a, b;
+            if (parentType === "VariableDeclarator") {
+                if (parent.parent.declarations.length === 1) {
+
+                    // prevent fix of variable in forOf and forIn loops.
+                    if (isLoop(parent.parent.parent) && parent.parent.parent.body !== parent.parent) {
+                        return null;
+                    }
+
+                    // removes only variable not semicolon in 'if (foo()) var bar;' or in 'loops' or in 'with' statement.
+                    if (
+                        parent.parent.parent.type === "IfStatement" ||
+                        isLoop(parent.parent.parent) ||
+                        (parent.parent.parent.type === "WithStatement" && parent.parent.parent.body === parent.parent)
+                    ) {
+                        return fixer.replaceText(parent.parent, ";");
+                    }
+
+                    // if next token is a string it could become a directive if node is removed -> no suggestion.
+                    const nextToken = sourceCode.getTokenAfter(parent.parent);
+
+                    if (nextToken) {
+                        if (nextToken.type === "String") {
+                            return null;
+                        }
+
+                        // if previous token exists and is not ";" or "{" not sure about ASI rules -> no suggestion.
+                        const prevToken = sourceCode.getTokenBefore(parent.parent);
+
+                        if (prevToken && !astUtils.isSemicolonToken(prevToken) && !astUtils.isOpeningBraceToken(prevToken)) {
+                            return null;
+                        }
+                    }
```

Comment:
This block is repeated in `fixVariables` and `fixObjectAndArrayVariable`. Maybe we could extract the logic into a new function and leave here just a simple check like:

```js
if (!isDeclarationSafeToRemove(parent, nextToken)) {
    return null;
}
```